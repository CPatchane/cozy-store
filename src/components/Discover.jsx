import React, { Component } from 'react'

import { translate } from 'cozy-ui/react/I18n'

import SmallAppItem from './SmallAppItem'

const DEMO_APPS = [
  {
    slug: 'bank',
    name: 'Bank',
    developer: {name: 'Cozy'},
    editor: 'Cozy',
    icon: require('../assets/temp_apps/icon-bank.svg')
  },
  {
    slug: 'sante',
    name: 'Sant√©',
    developer: {name: 'MAIF Ltd'},
    editor: '',
    icon: require('../assets/temp_apps/icon-sante.svg')
  },
  {
    slug: 'mon_logis',
    name: 'Mon Logis',
    developer: {name: 'HoodBrains'},
    editor: '',
    icon: require('../assets/temp_apps/icon-mon_logis.svg')
  },
  {
    slug: 'calendar',
    name: 'Calendar',
    developer: {name: 'Cozy'},
    editor: 'Cozy',
    icon: require('../assets/temp_apps/icon-calendar.svg')
  }
]

class Discover extends Component {
  render () {
    const { t } = this.props
    return (
      <div className='sto-discover sto-soon'>
        <h2 className='sto-discover-title'>{t('discover.title')}</h2>
        <div className='sto-discover-get-started'>
          <h3 className='sto-discover-get-started-title'>
            {t('discover.get_started')}
          </h3>
          <div className='sto-discover-get-started-list'>
            {DEMO_APPS.map(a => {
              return <SmallAppItem
                slug={a.slug}
                developer={a.developer}
                editor={a.editor}
                icon={a.icon}
                name={a.name}
                version={a.version}
              />
            })}
          </div>
        </div>
      </div>
    )
  }
}

export default translate()(Discover)
