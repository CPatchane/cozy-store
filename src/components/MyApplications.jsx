import React, { Component } from 'react'

import { Route } from 'react-router-dom'
import { translate } from 'cozy-ui/react/I18n'
import Spinner from 'cozy-ui/react/Spinner'

import SmallAppItem from './SmallAppItem'
import ApplicationModal from './ApplicationModal'

class MyApplications extends Component {
  constructor (props) {
    super(props)
    props.fetchApps()
  }

  onAppClick (appSlug) {
    this.props.history.push(`/myapps/${appSlug}/manage`)
  }

  render () {
    const { t, myApps, isFetching, error } = this.props
    return (
      <div className='sto-myapps'>
        <h2 className='sto-myapps-title'>{t('myapps.title')}</h2>
        <div className='sto-myapps-list'>
          {!isFetching && myApps && !!myApps.length &&
            myApps.map(a => {
              return <SmallAppItem
                slug={a.slug}
                developer={a.developer}
                editor={a.editor}
                icon={a.icon}
                name={a.name}
                version={a.version}
                onClick={() => this.onAppClick(a.slug)}
              />
            })
          }
          {error &&
            <p>{error}</p>
          }
        </div>
        {isFetching &&
          <Spinner
            size='xxlarge'
            loadingType='appsFetching'
            middle='true'
          />
        }

        <Route path='/myapps/:appSlug/manage' render={({ match }) =>
          (<ApplicationModal {...this.props} match={match} />)
        } />
      </div>
    )
  }
}

export default translate()(MyApplications)
