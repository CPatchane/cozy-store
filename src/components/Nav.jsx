import React from 'react'
import { translate } from 'cozy-ui/react/I18n'
import { NavLink } from 'react-router-dom'

export const Nav = ({ t }) => (
  <nav>
    <ul className='coz-nav'>
      <li className='coz-nav-item'>
        <NavLink to='/discover' className='sto-cat-discover coz-nav-link' activeClassName='active'>
          { t('nav.discover') }
        </NavLink>
      </li>
      <li className='coz-nav-item'>
        <NavLink to='/myapps' className='sto-cat-myapps coz-nav-link' activeClassName='active'>
          { t('nav.myapps') }
        </NavLink>
      </li>
    </ul>
  </nav>
)

export default translate()(Nav)
