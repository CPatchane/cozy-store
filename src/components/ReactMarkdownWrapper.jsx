import React from 'react'
import ReactMarkdown from 'react-markdown'

export const reactMarkdownRendererOptions = {
  Link: props => <a href={props.href} target='_blank'>{props.children}</a>
}

export const ReactMarkdownWrapper = ({ source }) =>
  <ReactMarkdown
    source={source}
    renderers={reactMarkdownRendererOptions}
  />

export default ReactMarkdownWrapper
