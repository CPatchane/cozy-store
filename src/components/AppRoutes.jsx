import React from 'react'
import { Route, Switch, Redirect } from 'react-router-dom'

import MyApplications from '../containers/MyApplications'

const AppRoutes = () => {
  return <Switch>
    <Route path='/myapps' component={MyApplications} />
    <Redirect from='/' to='/myapps' />
  </Switch>
}

export default AppRoutes
