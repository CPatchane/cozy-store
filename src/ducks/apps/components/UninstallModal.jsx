/* global cozy */
import React, { Component } from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'
import PropTypes from 'prop-types'

import { translate } from 'cozy-ui/react/I18n'
import Alerter from 'cozy-ui/react/Alerter'
import Button from 'cozy-ui/react/Button'
import { getAppBySlug, uninstallApp } from 'ducks/apps'
import Modal, { ModalContent } from 'cozy-ui/react/Modal'

import ReactMarkdownWrapper from '../../components/ReactMarkdownWrapper'

export class UninstallModal extends Component {
  constructor(props) {
    super(props)

    const { app, onNotInstalled } = props
    if (app && !app.installed) {
      onNotInstalled()
    }
  }

  componentDidUpdate = prevProps => {
    const { app, onSuccess, t } = this.props
    const haveJustBeenUninstalled = prevProps.app.installed && !app.installed
    if (haveJustBeenUninstalled) {
      Alerter.success(t('app_modal.uninstall.message.success'), {
        duration: 3000
      })
      onSuccess()
    }
  }

  uninstallApp = () => {
    const { app, uninstallApp } = this.props
    uninstallApp(app)
  }

  render() {
    const { isUninstalling, dismissAction, t, uninstallError } = this.props
    return (
      <div className="sto-modal--uninstall">
        <Modal
          title={t('app_modal.uninstall.title')}
          dismissAction={dismissAction}
          mobileFullscreen
        >
          <ModalContent>
            <div className="sto-modal-content">
              <ReactMarkdownWrapper
                source={t('app_modal.uninstall.description', {
                  cozyName: cozy.client._url.replace(/^\/\//, '')
                })}
              />
              {uninstallError && (
                <p className="u-error">
                  {t('app_modal.uninstall.message.error', {
                    message: uninstallError.message
                  })}
                </p>
              )}
              <div className="sto-modal-controls">
                <button
                  role="button"
                  className="c-btn c-btn--secondary"
                  onClick={dismissAction}
                >
                  <span>{t('app_modal.uninstall.cancel')}</span>
                </button>
                <Button
                  busy={isUninstalling}
                  disabled={isUninstalling}
                  theme="danger"
                  icon="delete"
                  onClick={this.uninstallApp}
                  label={t('app_modal.uninstall.uninstall')}
                />
              </div>
            </div>
          </ModalContent>
        </Modal>
      </div>
    )
  }
}

UninstallModal.propTypes = {
  app: PropTypes.object.isRequired,
  dismissAction: PropTypes.func.isRequired,
  onNotInstalled: PropTypes.func.isRequired,
  onSuccess: PropTypes.func.isRequired
}

const mapStateToProps = (state, ownProps) => ({
  app: getAppBySlug(state, ownProps.appSlug),
  isUninstalling: state.apps.isUninstalling,
  uninstallError: state.apps.actionError
})

const mapDispatchToProps = dispatch => ({
  uninstallApp: app => {
    return dispatch(uninstallApp(app))
  }
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(translate()(withRouter(UninstallModal)))
