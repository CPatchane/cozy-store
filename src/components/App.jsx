
import React from 'react'

import Sidebar from './Sidebar'
import AppRoutes from './AppRoutes'

const App = () => (
  <div className='sto-wrapper coz-sticky'>
    <Sidebar />
    <main className='sto-content'>
      <AppRoutes />
    </main>
  </div>
)
export default App
