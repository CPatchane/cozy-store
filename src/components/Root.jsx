import React from 'react'
import { Provider } from 'react-redux'
import { I18n } from 'cozy-ui/react/I18n'
import { HashRouter } from 'react-router-dom'

import App from './App'

const Root = ({ context, lang, store }) => {
  return <I18n lang={lang} dictRequire={(lang) => require(`../locales/${lang}`)}>
    <Provider store={store}>
      <HashRouter>
        <App />
      </HashRouter>
    </Provider>
  </I18n>
}

export default Root
