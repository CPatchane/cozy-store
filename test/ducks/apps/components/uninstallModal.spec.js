'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import { tMock } from '../../../jestLib/I18n'
import { UninstallModal } from 'ducks/apps/components/UninstallModal'

import mockApps from '../_mockApps'

/* SinonJS is used here to stub Promise in order to be synchronous.
In this way, (p)React will call setState synchronously. It will allow
to assert the component state juste after */
import sinon from 'sinon'
import sinonStubPromise from 'sinon-stub-promise'
sinonStubPromise(sinon)

const mockError = new Error('This is a test error')

const getMockProps = (slug, uninstallError = null) => ({
  app: mockApps.find(a => a.slug === slug),
  parent: '/myapps',
  uninstallApp: jest.fn(appSlug => {
    if (['drive', 'collect'].includes(appSlug)) {
      return sinon
        .stub()
        .returnsPromise()
        .rejects(mockError)()
    }
    return sinon
      .stub()
      .returnsPromise()
      .resolves({})()
  }),
  history: {
    push: jest.fn()
  },
  onNotInstalled: jest.fn(),
  uninstallError
})

describe('UninstallModal component', () => {
  beforeAll(() => {
    // define global mock url
    global.cozy = {
      client: {
        _url: '//cozytest.mock.cc'
      }
    }
  })

  it('should be rendered correctly (app uninstallable)', () => {
    const component = shallow(
      <UninstallModal t={tMock} {...getMockProps('photos')} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('calls onNotInstalled', () => {
    const props = getMockProps('photos')
    props.installed = false
    shallow(<UninstallModal t={tMock} {...props} />)
    expect(props.onNotInstalled.mock.calls.length).toBe(1)
  })

  it('should handle correctly error from props', () => {
    const mockProps = getMockProps('photos', mockError)
    const component = shallow(<UninstallModal t={tMock} {...mockProps} />)
    expect(component.getElement()).toMatchSnapshot()
  })
})
