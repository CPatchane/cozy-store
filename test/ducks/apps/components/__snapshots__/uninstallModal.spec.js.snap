// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UninstallModal component should be rendered correctly if app not uninstallable 1`] = `
<div
  className="sto-modal--uninstall"
>
  <Unknown
    secondaryAction={[Function]}
    title="Uninstall this application"
  >
    <ModalContent>
      <div
        className="sto-modal-content"
      >
        <ReactMarkdownWrapper
          source="This application is a major part of the Cozy platform. In order to keep you a __minimum user experience quality__, it's not possible to uninstall it through the Store application."
        />
        <div
          className="sto-modal-controls"
        >
          <button
            className="c-btn c-btn--secondary"
            onClick={[Function]}
            role="button"
          >
            Cancel
          </button>
          <button
            className="c-btn c-btn--danger c-btn--delete"
            disabled={true}
            onClick={[Function]}
            role="button"
          >
            Uninstall
          </button>
        </div>
      </div>
    </ModalContent>
  </Unknown>
</div>
`;

exports[`UninstallModal component should be rendered correctly if app uninstallable 1`] = `
<div
  className="sto-modal--uninstall"
>
  <Unknown
    secondaryAction={[Function]}
    title="Uninstall this application"
  >
    <ModalContent>
      <div
        className="sto-modal-content"
      >
        <ReactMarkdownWrapper
          source="This application is bound to be removed from your Cozy cozytest.mock.cc. Are you sure to __remove this application definitively__?"
        />
        <div
          className="sto-modal-controls"
        >
          <button
            className="c-btn c-btn--secondary"
            onClick={[Function]}
            role="button"
          >
            Cancel
          </button>
          <button
            className="c-btn c-btn--danger c-btn--delete"
            disabled={false}
            onClick={[Function]}
            role="button"
          >
            Uninstall
          </button>
        </div>
      </div>
    </ModalContent>
  </Unknown>
</div>
`;

exports[`UninstallModal component should hanlde correctly error from props 1`] = `
<div
  className="sto-modal--uninstall"
>
  <Unknown
    secondaryAction={[Function]}
    title="Uninstall this application"
  >
    <ModalContent>
      <div
        className="sto-modal-content"
      >
        <ReactMarkdownWrapper
          source="This application is bound to be removed from your Cozy cozytest.mock.cc. Are you sure to __remove this application definitively__?"
        />
        <p
          class="u-error"
        >
          Something went wrong when uninstalling this application. Reason: This is a test error
        </p>
        <div
          className="sto-modal-controls"
        >
          <button
            className="c-btn c-btn--secondary"
            onClick={[Function]}
            role="button"
          >
            Cancel
          </button>
          <button
            className="c-btn c-btn--danger c-btn--delete"
            disabled={false}
            onClick={[Function]}
            role="button"
          >
            Uninstall
          </button>
        </div>
      </div>
    </ModalContent>
  </Unknown>
</div>
`;
